server: TestEnsureSQLStatsAreFlushedForTelemetry failed

server.TestEnsureSQLStatsAreFlushedForTelemetry [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3113610&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3113610&tab=artifacts#/) on master @ [95028269160e0bfa2b94477238e72670bf780e9d](https://github.com/cockroachdb/cockroach/commits/95028269160e0bfa2b94477238e72670bf780e9d):


```
=== RUN   TestEnsureSQLStatsAreFlushedForTelemetry
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEnsureSQLStatsAreFlushedForTelemetry543294354
    test_log_scope.go:74: use -show-logs to present logs inline
    stats_test.go:152: expected to not find stat, but did
    stats_test.go:157: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEnsureSQLStatsAreFlushedForTelemetry543294354
--- FAIL: TestEnsureSQLStatsAreFlushedForTelemetry (0.22s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestEnsureSQLStatsAreFlushedForTelemetry PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
rohany marked as alumn{us/a}; resolving to solongordon instead
```
</p>
</details>

/cc @cockroachdb/server solongordon
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEnsureSQLStatsAreFlushedForTelemetry.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/server/stats_test.go
+++ b/pkg/server/stats_test.go
@@ -14,24 +14,21 @@ import (
 	"context"
 	gosql "database/sql"
 	"reflect"
-	"strings"
 	"testing"
-	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/server/diagnostics"
 	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
-	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/diagutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
-	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
 
@@ -104,24 +101,37 @@ func TestEnsureSQLStatsAreFlushedForTelemetry(t *testing.T) {
 	ctx := context.Background()
 	params, _ := tests.CreateTestServerParams()
 	params.Settings = cluster.MakeClusterSettings()
-	// Set the SQL stat refresh rate very low so that SQL stats are continuously
-	// flushed into the telemetry reporting stats pool.
-	sqlstats.SQLStatReset.Override(ctx, &params.Settings.SV, 10*time.Millisecond)
 	s, sqlDB, _ := serverutils.StartServer(t, params)
 	defer s.Stopper().Stop(ctx)
 
+	sqlConn := sqlutils.MakeSQLRunner(sqlDB)
+
+	tcs := []struct {
+		stmt        string
+		fingerprint string
+	}{
+		{
+			stmt:        "SELECT 1",
+			fingerprint: "SELECT _",
+		},
+		{
+			stmt:        "SELECT 1, 1",
+			fingerprint: "SELECT _, _",
+		},
+		{
+			stmt:        "SELECT 1, 1, 1",
+			fingerprint: "SELECT _, _, _",
+		},
+	}
+
 	// Run some queries against the database.
-	if _, err := sqlDB.Exec(`
-CREATE DATABASE t;
-CREATE TABLE t.test (x INT PRIMARY KEY);
-INSERT INTO t.test VALUES (1);
-INSERT INTO t.test VALUES (2);
-`); err != nil {
-		t.Fatal(err)
+	for _, tc := range tcs {
+		sqlConn.Exec(t, tc.stmt)
 	}
 
 	statusServer := s.(*TestServer).status
 	sqlServer := s.(*TestServer).Server.sqlServer.pgServer.SQLServer
+	sqlServer.GetSQLStatsController().ResetLocalSQLStats(ctx)
 	testutils.SucceedsSoon(t, func() error {
 		// Get the diagnostic info.
 		res, err := statusServer.Diagnostics(ctx, &serverpb.DiagnosticsRequest{NodeId: "local"})
@@ -129,17 +139,18 @@ INSERT INTO t.test VALUES (2);
 			t.Fatal(err)
 		}
 
-		found := false
+		foundFingerprintCnt := 0
 		for _, stat := range res.SqlStats {
 			// These stats are scrubbed, so look for our scrubbed statement.
-			if strings.HasPrefix(stat.Key.Query, "INSERT INTO _ VALUES (_)") {
-				found = true
+			for _, tc := range tcs {
+				if tc.fingerprint == stat.Key.Query {
+					foundFingerprintCnt++
+					break
+				}
 			}
 		}
 
-		if !found {
-			return errors.New("expected to find query stats, but didn't")
-		}
+		require.Equal(t, len(tcs), foundFingerprintCnt, "expected to find query stats, but didn't")
 
 		// We should also not find the stat in the SQL stats pool, since the SQL
 		// stats are getting flushed.
@@ -147,9 +158,8 @@ INSERT INTO t.test VALUES (2);
 		require.NoError(t, err)
 
 		for _, stat := range stats {
-			// These stats are scrubbed, so look for our scrubbed statement.
-			if strings.HasPrefix(stat.Key.Query, "INSERT INTO _ VALUES (_)") {
-				t.Error("expected to not find stat, but did")
+			for _, tc := range tcs {
+				require.NotEqual(t, tc.fingerprint, stat.Key.Query, "expected to not found %s in stats, bud did", stat.Key.Query)
 			}
 		}
 		return nil

